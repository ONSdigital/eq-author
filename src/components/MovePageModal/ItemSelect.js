import React from "react";
import styled from "styled-components";
import { colors } from "constants/theme";
import PropTypes from "prop-types";
import VisuallyHidden from "components/VisuallyHidden";
import iconMoveIndicator from "./icon-move-indicator.svg";
import { uniqueId } from "lodash";
import withChangeHandler from "../Forms/withChangeHandler";

const Input = VisuallyHidden.withComponent("input");
Input.defaultProps = {
  type: "radio"
};

const Truncated = styled.span`
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Label = styled.label`
  display: flex;
  align-items: center;
  padding: 0.5em 0.5em 0.5em 2.25em;
  position: relative;

  &:hover {
    background-color: ${colors.lighterGrey};
  }

  ${/* sc-sel */ Input}:focus + & {
    outline: 2px solid ${colors.text};
    outline-offset: -2px;
  }

  ${/* sc-sel */ Input}:checked + & {
    background-color: ${colors.blue};
    color: ${colors.white};
  }

  &:hover::before {
    background-color: ${colors.text};
  }

  &:hover::before,
  ${/* sc-sel */ Input}:checked + &::before {
    content: "";
    position: absolute;
    left: 1rem;
    width: 1em;
    height: 1em;
    mask-image: url('${iconMoveIndicator}');
    mask-repeat: no-repeat;
    mask-position: center center;
  }

  ${/* sc-sel */ Input}:checked + &::before {
    background-color: white;
  }
`;

export const Option = ({
  name,
  selected,
  value,
  onChange,
  id = uniqueId("ItemList_Option"),
  children,
  ...otherProps
}) => (
  <div {...otherProps}>
    <Input
      value={value}
      id={id}
      onChange={onChange}
      checked={selected}
      name={name}
    />

    <Label selected={selected} htmlFor={id}>
      <Truncated>{children}</Truncated>
    </Label>
  </div>
);

Option.propTypes = {
  selected: PropTypes.bool,
  value: PropTypes.string.isRequired,
  name: PropTypes.string,
  onChange: PropTypes.func,
  id: PropTypes.string,
  children: PropTypes.node.isRequired
};

const ItemSelect = ({ children, value, name, onChange, ...otherProps }) => (
  <div {...otherProps}>
    {React.Children.map(children, child =>
      React.cloneElement(child, {
        onChange,
        name,
        selected: child.props.value === value
      })
    )}
  </div>
);

ItemSelect.propTypes = {
  children: PropTypes.node.isRequired,
  value: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
};

export default withChangeHandler(ItemSelect);
