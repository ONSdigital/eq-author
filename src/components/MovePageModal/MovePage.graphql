#import "graphql/fragments/movePage.graphql"

query MovePage($id: ID!) {
  questionnaire(id: $id) {
    sections {
      ...MovePageFragment
    }
  }
}
