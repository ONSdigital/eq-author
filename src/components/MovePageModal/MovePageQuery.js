import React from "react";
import { Query } from "react-apollo";
import query from "./MovePage.graphql";
import PropTypes from "prop-types";

const MovePageQuery = ({ questionnaireId, children }) => (
  <Query query={query} variables={{ id: questionnaireId }}>
    {({ loading, data }) => {
      return children({ loading, data });
    }}
  </Query>
);

MovePageQuery.propTypes = {
  questionnaireId: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired
};

export default MovePageQuery;
