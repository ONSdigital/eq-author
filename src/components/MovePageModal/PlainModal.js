import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Modal from "components/Modal";
import ScrollPane from "components/ScrollPane";

const Fieldset = styled.fieldset`
  border: none;
  padding: 0;
  min-width: 0;
`;

const Title = styled.legend`
  margin: 0;
  padding: 0.75rem;
  color: #858585;
  font-size: 1.25em;
  font-weight: bold;
  display: block;
`;

const BodyWrapper = styled.div`
  max-height: 18em;
`;

const Body = ({ children }) => (
  <ScrollPane>
    <BodyWrapper>{children}</BodyWrapper>
  </ScrollPane>
);

Body.propTypes = {
  children: PropTypes.node.isRequired
};

const PlainModal = styled(Modal).attrs({ hasCloseButton: false })`
  .Overlay {
    background: transparent;
  }

  .Modal {
    width: 23em;
    padding: 0;
    top: 8em;
  }
`;

export default PlainModal;
export { Fieldset, Title, Body };
