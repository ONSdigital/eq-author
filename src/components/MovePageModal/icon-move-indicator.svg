
<svg width="14px" height="14px" viewBox="0 0 14 14" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 49 (51002) - http://www.bohemiancoding.com/sketch -->
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Symbols" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Icon/Move-indicator-highlight-Copy" transform="translate(-3.000000, -3.000000)">
            <g id="Group">
                <rect id="Rectangle-4" x="0" y="0" width="20" height="20"></rect>
                <polygon id="î ®" fill="#666666" points="10 3.32 16.68 10 10 16.68 8.82 15.5 13.48 10.82 3.32 10.82 3.32 9.18 13.48 9.18 8.82 4.5"></polygon>
            </g>
        </g>
    </g>
</svg>