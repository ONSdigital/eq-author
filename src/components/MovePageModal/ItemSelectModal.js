import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Button from "components/Button";
import PlainModal from "./PlainModal";
import ButtonGroup from "components/ButtonGroup";
import ScrollPane from "components/ScrollPane";

const Fieldset = styled.fieldset`
  border: none;
  padding: 0;
  min-width: 0;
`;

const Title = styled.legend`
  margin: 0;
  padding: 0.75rem;
  color: #858585;
  font-size: 1.25em;
  font-weight: bold;
  display: block;
`;

const BodyWrapper = styled.div`
  max-height: 18em;
`;

const Body = props => (
  <ScrollPane>
    <BodyWrapper {...props} />
  </ScrollPane>
);

const StyledButtonGroup = styled(ButtonGroup).attrs({
  horizontal: true
})`
  padding: 1em;
  justify-content: space-around;
  border-top: 1px solid #aeaeae;
`;

const StyledButton = styled(Button)`
  flex: 1;
`;

const ItemSelectModal = ({
  isOpen,
  onClose,
  onConfirm,
  title,
  secondaryText,
  primaryText,
  children,
  ...otherProps
}) => (
  <PlainModal isOpen={isOpen} onClose={onClose} {...otherProps}>
    <form onSubmit={onConfirm}>
      <Fieldset>
        <Title>{title}</Title>
        <Body>{children}</Body>
      </Fieldset>
      <StyledButtonGroup>
        <StyledButton onClick={onClose} secondary type="button">
          {secondaryText}
        </StyledButton>
        <StyledButton type="submit" primary>
          {primaryText}
        </StyledButton>
      </StyledButtonGroup>
    </form>
  </PlainModal>
);

ItemSelectModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  primaryText: PropTypes.string,
  secondaryText: PropTypes.string,
  title: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired
};

ItemSelectModal.defaultProps = {
  secondaryText: "Cancel",
  primaryText: "Select"
};

export default ItemSelectModal;
