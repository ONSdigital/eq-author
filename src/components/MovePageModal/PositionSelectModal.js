import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import styled from "styled-components";
import Button from "components/Button";
import ItemSelect, { Option } from "./ItemSelect";
import PlainModal, { Title, Body, Fieldset } from "./PlainModal";
import ButtonGroup from "components/ButtonGroup";
import getTextFromHTML from "utils/getTextFromHTML";

const StyledButtonGroup = styled(ButtonGroup).attrs({
  horizontal: true,
  align: "right"
})`
  padding: 1em;
  justify-content: space-around;
  border-top: 1px solid #aeaeae;
`;

const StyledButton = styled(Button)`
  flex: 1;
`;

const PositionSelectModal = ({
  isOpen,
  onClose,
  onChange,
  selectedPosition,
  pages,
  onConfirm
}) => (
  <PlainModal isOpen={isOpen} onClose={onClose}>
    <form onSubmit={onConfirm}>
      <Fieldset>
        <Title>Position</Title>
        <Body>
          <ItemSelect
            value={String(selectedPosition)}
            onChange={onChange}
            name="position"
          >
            {pages.map((page, i) => (
              <Option key={i} value={String(i)}>
                {getTextFromHTML(page.title)}
              </Option>
            ))}
          </ItemSelect>
        </Body>
      </Fieldset>
      <StyledButtonGroup>
        <StyledButton onClick={onClose} secondary type="button">
          Cancel
        </StyledButton>
        <StyledButton type="submit" primary>
          Move
        </StyledButton>
      </StyledButtonGroup>
    </form>
  </PlainModal>
);

PositionSelectModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  selectedPosition: PropTypes.number.isRequired,
  pages: PropTypes.arrayOf(CustomPropTypes.page).isRequired,
  onConfirm: PropTypes.func.isRequired
};

export default PositionSelectModal;
