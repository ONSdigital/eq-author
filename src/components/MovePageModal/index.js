import React from "react";
import Modal from "components/Modal";
import styled from "styled-components";
import DialogHeader from "components/Dialog/DialogHeader";
import { Message, Heading } from "components/Dialog/DialogMessage";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import SectionSelectModal from "./SectionSelectModal";
import PositionSelectModal from "./PositionSelectModal";
import { find, reject, parseInt } from "lodash";
import getTextFromHTML from "utils/getTextFromHTML";
import Icon from "assets/icon-select.svg";

const StyledModal = styled(Modal)`
  .Modal {
    width: 25em;
  }
`;

const moveTo = (array, item, position) => {
  array = reject(array, item);
  array.splice(position, 0, item);

  return array;
};

const CenteredHeading = styled(Heading)`
  text-align: center;
  margin-bottom: 1rem;
`;

const Label = styled.p`
  font-size: 0.875em;
  font-weight: bold;
  margin-bottom: 0.25rem;
  margin-top: 1.25rem;
`;

const Trigger = styled.button.attrs({ type: "button" })`
  width: 100%;
  font-size: 0.875em;
  padding: 0.5rem;
  background: white url('${Icon}') no-repeat right 0.5em center;
  border: solid 1px #aeaeae;
  text-align: left;
  border-radius: 3px;
`;

class MovePageModal extends React.Component {
  static propTypes = {
    section: CustomPropTypes.section,
    page: CustomPropTypes.page,
    questionnaire: CustomPropTypes.questionnaire.isRequired,
    isOpen: PropTypes.bool.isRequired,
    onClose: PropTypes.func.isRequired,
    onMovePage: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      isSectionSelectOpen: false,
      isPagePositionOpen: false,
      selectedSectionId: props.section.id,
      selectedPagePosition: props.page.position
    };
  }

  handleToggleSectionSelect = () => {
    this.setState({ isSectionSelectOpen: !this.state.isSectionSelectOpen });
  };

  handleTogglePagePosition = () => {
    this.setState({ isPagePositionOpen: !this.state.isPagePositionOpen });
  };

  handleSectionSelect = ({ value }) => {
    this.setState({
      selectedSectionId: value,
      selectedPagePosition: 0,
      isSectionSelectOpen: false
    });
  };

  handlePositionSelect = ({ value }) => {
    this.setState({ selectedPagePosition: parseInt(value) });
  };

  handlePageMove = e => {
    const { page, onMovePage, onClose } = this.props;
    const { selectedSectionId, selectedPagePosition } = this.state;

    e.preventDefault();

    this.setState({ isPagePositionOpen: false }, () => {
      onMovePage({
        id: page.id,
        sectionId: selectedSectionId,
        position: selectedPagePosition
      });
      onClose();
    });
  };

  getSection() {
    const { questionnaire } = this.props;
    const { selectedSectionId } = this.state;

    return find(questionnaire.sections, { id: selectedSectionId });
  }

  getPages(section) {
    return moveTo(
      section.pages,
      this.props.page,
      this.state.selectedPagePosition
    );
  }

  render() {
    const { isOpen, onClose, questionnaire } = this.props;
    const {
      isPagePositionOpen,
      isSectionSelectOpen,
      selectedPagePosition
    } = this.state;

    const section = this.getSection();
    const pages = this.getPages(section);

    return (
      <StyledModal isOpen={isOpen} onClose={onClose}>
        <DialogHeader>
          <Message>
            <CenteredHeading>Move question</CenteredHeading>
          </Message>
        </DialogHeader>
        <Label>Section</Label>
        <Trigger onClick={this.handleToggleSectionSelect}>
          {getTextFromHTML(section.title)}
        </Trigger>
        <SectionSelectModal
          isOpen={isSectionSelectOpen}
          onClose={this.handleToggleSectionSelect}
          onChange={this.handleSectionSelect}
          sections={questionnaire.sections}
          selectedSection={section}
        />
        <Label>Position</Label>
        <Trigger onClick={this.handleTogglePagePosition}>Select</Trigger>
        <PositionSelectModal
          isOpen={isPagePositionOpen}
          onClose={this.handleTogglePagePosition}
          onChange={this.handlePositionSelect}
          onConfirm={this.handlePageMove}
          pages={pages}
          selectedPosition={selectedPagePosition}
        />
      </StyledModal>
    );
  }
}

export default MovePageModal;
