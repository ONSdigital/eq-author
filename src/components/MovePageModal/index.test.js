import React from "react";
import MovePageModal from ".";
import { shallow } from "enzyme";
import { times } from "lodash";

const buildPages = (sectionNumber, count) =>
  times(count, i => ({
    id: `${sectionNumber}.${i + 1}`,
    title: `Page ${sectionNumber}.${i + 1}`,
    position: i
  }));

const buildSections = count =>
  times(count, i => ({
    id: `${i + 1}`,
    title: `Section ${i + 1}`,
    pages: buildPages(i + 1, 2)
  }));

const buildQuestionnaire = () => ({
  id: "1",
  sections: buildSections(2)
});

const sel = id => `[data-test="${id}"]`;
const getSectionModal = wrapper => wrapper.find(sel("section-modal"));
const getSectionSelect = wrapper => wrapper.find(sel("section-select"));
const getPositionModal = wrapper => wrapper.find(sel("position-modal"));
const getPositionSelect = wrapper => wrapper.find(sel("position-select"));

describe("MovePageModal/MovePageModal", () => {
  const questionnaire = buildQuestionnaire();
  const currentSection = questionnaire.sections[0];
  const currentPage = currentSection.pages[0];

  const createWrapper = (props = {}, render = shallow) =>
    render(
      <MovePageModal
        isOpen
        onClose={jest.fn()}
        questionnaire={questionnaire}
        section={currentSection}
        page={currentPage}
        selectedPosition={0}
        onMovePage={jest.fn()}
        {...props}
      />
    );

  it("should render", () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it("opens section select modal when correct button is clicked", () => {
    const wrapper = createWrapper();

    wrapper
      .find("MovePageModal__Trigger")
      .first()
      .simulate("click");

    expect(getSectionModal(wrapper).prop("isOpen")).toBe(true);
  });

  it("opens position select modal when correct button is clicked", () => {
    const wrapper = createWrapper();

    wrapper
      .find("MovePageModal__Trigger")
      .last()
      .simulate("click");

    expect(getPositionModal(wrapper).prop("isOpen")).toBe(true);
  });

  it("should close section select modal on confirm", () => {
    const wrapper = createWrapper();

    wrapper
      .find("MovePageModal__Trigger")
      .first()
      .simulate("click");

    getSectionModal(wrapper).simulate("change", {
      value: questionnaire.sections[1].id
    });

    getSectionModal(wrapper).simulate("confirm", {
      preventDefault: jest.fn()
    });

    expect(getSectionModal(wrapper).prop("isOpen")).toBe(false);
  });

  it("should update selected section on change", () => {
    const wrapper = createWrapper();
    const selectedSection = questionnaire.sections[1];

    getSectionSelect(wrapper).simulate("change", { value: selectedSection.id });

    expect(getSectionSelect(wrapper).prop("value")).toBe(selectedSection.id);
  });

  it("should update selected position on change", () => {
    const wrapper = createWrapper();

    getSectionSelect(wrapper).simulate("change", { value: "2" });
    getPositionSelect(wrapper).simulate("change", { value: "1" });

    expect(getPositionSelect(wrapper)).toMatchSnapshot();
  });

  it("closes all modals and calls onMovePage when confirmed", () => {
    const onMovePage = jest.fn();
    const onClose = jest.fn();
    const selectedSection = questionnaire.sections[1];
    const position = 1;

    const wrapper = createWrapper({ onMovePage, onClose });

    getSectionSelect(wrapper).simulate("change", { value: selectedSection.id });
    getPositionSelect(wrapper).simulate("change", { value: String(position) });
    getPositionModal(wrapper).simulate("confirm", {
      preventDefault: jest.fn()
    });

    expect(getPositionModal(wrapper).prop("isOpen")).toBe(false);
    expect(onClose).toHaveBeenCalled();

    expect(onMovePage).toHaveBeenCalledWith({
      id: currentPage.id,
      sectionId: selectedSection.id,
      position
    });
  });
});
