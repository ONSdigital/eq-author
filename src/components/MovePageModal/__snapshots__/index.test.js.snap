// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovePageModal/MovePageModal should render 1`] = `
<MovePageModal__StyledModal
  isOpen={true}
  onClose={[Function]}
>
  <DialogHeader___default>
    <DialogMessage__Message>
      <MovePageModal__CenteredHeading>
        Move question
      </MovePageModal__CenteredHeading>
    </DialogMessage__Message>
  </DialogHeader___default>
  <MovePageModal__Label>
    Section
  </MovePageModal__Label>
  <MovePageModal__Trigger
    onClick={[Function]}
  >
    Section 1
  </MovePageModal__Trigger>
  <ItemSelectModal
    data-test="section-modal"
    isOpen={false}
    onClose={[Function]}
    onConfirm={[Function]}
    primaryText="Select"
    secondaryText="Cancel"
    title="Section"
  >
    <withChangeHandler(ItemSelect)
      data-test="section-select"
      name="section"
      onChange={[Function]}
      value="1"
    >
      <Option
        key="1"
        value="1"
      >
        Section 1
      </Option>
      <Option
        key="2"
        value="2"
      >
        Section 2
      </Option>
    </withChangeHandler(ItemSelect)>
  </ItemSelectModal>
  <MovePageModal__Label>
    Position
  </MovePageModal__Label>
  <MovePageModal__Trigger
    onClick={[Function]}
  >
    Select
  </MovePageModal__Trigger>
  <ItemSelectModal
    data-test="position-modal"
    isOpen={false}
    onClose={[Function]}
    onConfirm={[Function]}
    primaryText="Move page"
    secondaryText="Cancel"
    title="Position"
  >
    <withChangeHandler(ItemSelect)
      data-test="position-select"
      name="position"
      onChange={[Function]}
      value="0"
    >
      <Option
        key="0"
        value="0"
      >
        Page 1.1
      </Option>
      <Option
        key="1"
        value="1"
      >
        Page 1.2
      </Option>
    </withChangeHandler(ItemSelect)>
  </ItemSelectModal>
</MovePageModal__StyledModal>
`;

exports[`MovePageModal/MovePageModal should update selected position on change 1`] = `
<withChangeHandler(ItemSelect)
  data-test="position-select"
  name="position"
  onChange={[Function]}
  value="1"
>
  <Option
    key="0"
    value="0"
  >
    Page 2.1
  </Option>
  <Option
    key="1"
    value="1"
  >
    Page 1.1
  </Option>
  <Option
    key="2"
    value="2"
  >
    Page 2.2
  </Option>
</withChangeHandler(ItemSelect)>
`;
